### PR TITLE
Allow member fields from other operations in spec

### PR DESCRIPTION
**Pull request by RedbackThomson**
Description of changes:
Allow the use of member fields contained within other API operations within a resource's spec. For example, `LoggingStatus` for S3 is only in the `PutBucketLogging` operation, but should be settable from spec.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by vijtrip2** [comment]
Is this a PR in progress? 
From the code changes, I could not relate how these change will allow mentioning spec fields from another operation.  Can you please elaborate for me? 
Also No unit tests added?

**Comment by RedbackThomson** [comment]
> Is this a PR in progress?
Nope. This is just a small PR.

> From the code changes, I could not relate how these change will allow mentioning spec fields from another operation. Can you please elaborate for me?
I want to add `LoggingStatus` as a `spec` field for S3. To do this, I would add a new field in the configuration and set the operation to `PutBucketLogging` with path `LoggingStatus`. When I go to generate this code, the generator will add the `LoggingStatus` field to the CRD, but the underlying go type for that field (and any of its members) are not added to the type defs. This is because [this part of the generator](https://github.com/aws-controllers-k8s/code-generator/blob/5a40e37032506f12a07b7de12a6005b525d13ad9/pkg/model/model.go#L272-L280) ensures that all type defs are in the input or output of some API - but in this case `PutBucketLogging` is neither.

> Also No unit tests added?
I don't see any unit tests for the type defs methods. Any pointers?

**Comment by vijtrip2** [comment]
> > Is this a PR in progress?
> 
> Nope. This is just a small PR.
> 
> > From the code changes, I could not relate how these change will allow mentioning spec fields from another operation. Can you please elaborate for me?
> 
> I want to add `LoggingStatus` as a `spec` field for S3. To do this, I would add a new field in the configuration and set the operation to `PutBucketLogging` with path `LoggingStatus`. When I go to generate this code, the generator will add the `LoggingStatus` field to the CRD, but the underlying go type for that field (and any of its members) are not added to the type defs. This is because [this part of the generator](https://github.com/aws-controllers-k8s/code-generator/blob/5a40e37032506f12a07b7de12a6005b525d13ad9/pkg/model/model.go#L272-L280) ensures that all type defs are in the input or output of some API - but in this case `PutBucketLogging` is neither.
> 
> > Also No unit tests added?
> 
> I don't see any unit tests for the type defs methods. Any pointers?

Thanks for explaining. We should ideally make an effort to increase test coverage of our codebase (maybe a collective effort as a team in coming sprints) but it's definitely out of scope for this PR.

LGTM. I'll let one more team member review this before merging.

**Review by vijtrip2 — APPROVED** [review]

File: pkg/model/crd.go
Comment (jaypipes) [inline pkg/model/crd.go:254]: This isn't needed, AFAICT, nor is the `CRD.IsShapeInSpec` method below since the block of code you added above in `CRD.HasShapeAsMember` effectively will return true when the CRD has a shape as a member (or nested member). I believe you only added this as a caching mechanism in the GetTypeDefs() function, is that right?

I'd prefer if we kept the code as simple as possible and so think you can remove this and the `IsShapeInSpec` methods.

**Review by jaypipes — CHANGES_REQUESTED** [review]
@RedbackThomson please see my inline comment.

Also, to unit test this functionality, simply add the generator config changes you are testing out for the S3 controller's update code paths to the testdata/ generator yaml for S3:

https://github.com/aws-controllers-k8s/code-generator/blob/5a40e37032506f12a07b7de12a6005b525d13ad9/pkg/testdata/models/apis/s3/0000-00-00/generator.yaml

and add a test for the now-expected new Spec field for Logging in this unit test file:

https://github.com/aws-controllers-k8s/code-generator/blob/5a40e37032506f12a07b7de12a6005b525d13ad9/pkg/model/model_s3_test.go#L70

and a test for the now-expected LoggingStatus type defs in the same model_s3_test.go file. You can use the `testutil.GetTypeDefByName` utility function from here:

https://github.com/aws-controllers-k8s/code-generator/blob/5a40e37032506f12a07b7de12a6005b525d13ad9/pkg/testutil/get.go#L45

File: pkg/model/crd.go
Comment (RedbackThomson) [reply pkg/model/crd.go:254]: Oh wow. You are absolutely correct. My change to `CRD.HasShapeAsMember` came after I added these other methods, and that is the only thing that was necessary. Thank you!

**Comment by RedbackThomson** [comment]
Cleaned up unnecessary methods and added unit tests for the S3 `Logging` type

**Review by A-Hilaly — APPROVED** [review]
👍

File: pkg/model/model_s3_test.go
Comment (jaypipes) [inline pkg/model/model_s3_test.go:100]: :+1:

**Review by jaypipes — APPROVED** [review]
Bingo. ++

**Pull request merged**
